whatsnew overhaul (#3769)
SciTools/iris: docs/iris/src/developers_guide/release.rst

Merge back
----------

After the release is cut, the changes shall be merged back onto the
Scitools/iris master branch.

To achieve this, first cut a local branch from the release branch,
:literal:`{release}.x`.  Next add a commit changing the release string to match
the release string on scitools/master.  This branch can now be proposed as a
pull request to master.  This work flow ensures that the commit identifiers are
consistent between the :literal:`.x` branch and :literal:`master`.

Point releases
--------------

Bug fixes may be implemented and targeted as the :literal:`.x` branch.  These
should lead to a new point release, another tag.  For example, a fix for a
problem with 1.9.0 will be merged into 1.9.x, and then released by tagging
1.9.1.

New features shall not be included in a point release, these are for bug fixes.

A point release does not require a release candidate, but the rest of the
release process is to be followed, including the merge back of changes into
:literal:`master`.  

.. _iris_development_releases_steps:

Maintainer steps
----------------

These steps assume a release for ``v1.9`` is to be created

Release steps
~~~~~~~~~~~~~

> #. Create a branch on the main repo, not in a forked repo, for the release

Reviewer: ```suggestion
#. Create the branch "1.9.x" on the main repo, not in a forked repo, for the release
   (Except for a point/bugfix release, when it already exists).
```

   candidate or release
#. Update the what's new for the release:  

    * Copy ``docs/iris/src/whatsnew/latest.rst`` to a file named
      ``v1.9.rst``
    * Delete the ``docs/iris/src/whatsnew/latest.rst`` file so it will not
      cause an issue in the build
    * In ``v1.9.rst`` update the page title (first line of the file) to show
      the date and version in the format of ``v1.9 (DD MMM YYYY)``.  For
      example ``v1.9 (03 Aug 2020)``
    * Review the file for correctness

>     * Add ``v1.9.rst`` and updates to ``latest.rst`` to git

Reviewer: ```suggestion
    * Add ``v1.9.rst`` to git and commit all changes, including removal of ``latest.rst``.
```

#. Update the what's new index ``docs/iris/src/whatsnew/index.rst``

>    * Temporary remove reference to ``latest.rst``

Reviewer: ```suggestion
   * Temporarily remove reference to ``latest.rst``
```

   * Add a reference to ``v1.9.rst`` to the top of the list

#. Update ``Iris.__init__.py`` version string
#. Check your changes by building the documentation and viewing the changes

> #. Once all the above steps are complete, the release is cut by the creation
>    of a new tag in the head of the release branch in the SciTools 

Reviewer: ```suggestion
#. Once all the above steps are complete, the release is cut, using the "Draft a new release" button
   on the `iris releases page <https://github.com/SciTools/iris/releases>`_ .
```

>    `Iris repository <https://github.com/tkknight/iris/releases>`_

Reviewer: This link should point to SciTools ?

Author: it should indeed, good spot.

Post release steps
~~~~~~~~~~~~~~~~~~

#. Check the documentation has built on `Read The Docs`_.  The build is 

>    triggered by any commit to master.  Additionallty check that the versions

Reviewer: ```suggestion
   triggered by any commit to master.  Additionally check that the versions
```

.. _iris_development_releases:

Releases
========

A release of Iris is a `tag on the SciTools/Iris`_ 
Github repository.

The summary below is of the main areas that constitute the release.  The final
section details the :ref:`iris_development_releases_steps` to take.

Release branch
--------------

Once the features intended for the release are on master, a release branch 
should be created, in the SciTools/Iris repository.  This will have the name:

>     :literal:`{major release number}.{minor release number}.x`

Reviewer 2: @tkknight You need to add the leading `v`

Author: yes i do

for example:

    :literal:`v1.9.x`

This branch shall be used to finalise the release details in preparation for
the release candidate.

Release candidate
-----------------

Prior to a release, a release candidate tag may be created, marked as a
pre-release in github, with a tag ending with :literal:`rc` followed by a
number, e.g.:

    :literal:`v1.9.0rc1`

> If created, the pre-release shall be available for at least one week prior to

Reviewer 2: @tkknight `one week` isn't a long time... don't know if we want to stipulate this here, but I'd think that an `rc` needs at least two weeks, if not four weeks, of availablity to the community for user testing

Author: will update to show 2 weeks, but ideally 4.

the release being cut.  User groups should be notified of the existence of the

> pre-release and encouraged to test the functionality.

Reviewer 2: @tkknight At a minimum we'd announce this on our scitools-iris twitter feed...

Author: this could be added - but this step would need to be added to the maintainer steps too.

A pre-release is expected for a minor release, but not normally provided for a

> point release.

Reviewer 2: @tkknight I'd be firmer than that, and state that there will not be any `rc`s for a point/patch release

Reviewer 2: That's actually stated later on... see [here](https://github.com/SciTools/iris/pull/3769/files#diff-357e85c5f819fb7f0efb53a02eeee65bR113)

If new features are required for a release after a release candidate has been
cut, a new pre-release shall be issued first.

Documentation
-------------

The documentation should include all of the what's new snippets, which must be
compiled into a what's new.  This content should be reviewed and adapted as
required and the snippets removed from the branch to produce a coherent what's

> new page.

Reviewer 2: @tkknight `coherent single what's new page`

Upon release, the documentation shall be added to the SciTools scitools.org.uk
github project's gh-pages branch as the latest documentation.

Testing the conda recipe
------------------------

Before a release is cut, the SciTools conda-recipes-scitools recipe for Iris
shall be tested to build the release branch of Iris; this test recipe shall not

> be merged onto conda-recipes-scitools.

Reviewer 2: @tkknight I don't see the point in this section now, and `SciTools conda-recipes-scitools` is well defunct now. Personally, I'd drop this section...

Author: Agreed.  Will remove.

The release
-----------

The final steps are to change the version string in the source of 
:literal:`Iris.__init__.py` and include the release date in the relevant what's
new page within the documentation.

Once all checks are complete, the release is cut by the creation of a new tag
in the SciTools Iris repository.

Conda recipe
------------

Once a release is cut, the `Iris feedstock`_ for the conda recipe must be
updated to build the latest release of Iris and push this artefact to
`conda forge`_.  

.. _Iris feedstock: https://github.com/conda-forge/iris-feedstock/tree/master/recipe
.. _conda forge: https://anaconda.org/conda-forge/iris

Merge back
----------

After the release is cut, the changes shall be merged back onto the scitools

> master.

Reviewer 2: @tkknight `onto the Scitools/iris master branch`

consistent between the :literal:`.x` branch and :literal:`master`.

Point releases
--------------

Bug fixes may be implemented and targeted as the :literal:`.x` branch.  These
should lead to a new point release, another tag.  For example, a fix for a
problem with 1.9.0 will be merged into 1.9.x, and then released by tagging
1.9.1.

New features shall not be included in a point release, these are for bug fixes.

A point release does not require a release candidate, but the rest of the
release process is to be followed, including the merge back of changes into
:literal:`master`.  

.. _iris_development_releases_steps:

Maintainer steps
----------------

These steps assume a release for ``v1.9`` is to be created

Release steps
~~~~~~~~~~~~~

#. Create a branch on the main repo, not in a forked repo, for the release
   candidate or release
#. Update the what's new for the release:  

    * Copy ``docs/iris/src/whatsnew/latest.rst`` to a file named
      ``v1.9.rst``
    * Delete the ``docs/iris/src/whatsnew/latest.rst`` file so it will not
      cause an issue in the build
    * In ``v1.9.rst`` update the page title (first line of the file) to show

>       the date and version in the format of ``v1.9 (DD MMM YYYY)``

Reviewer 2: @tkknight Perhaps give an example here for the date format to avoid confusion e.g., `01 Jan 1970`
